File: php/muse/data_structures/LinkedQueue.php
<?php

declare(strict_types=1);

namespace muse\data_structures\LinkedQueue;

require_once "muse/data_structures/interfaces/IQueue.php";

use muse\data_structures\interfaces\IQueue;

class Node
{
    public $data;
    public $next = null;

    public function __construct(int $element)
    {
        $this->data = $element;
    }
}

class LinkedQueue implements IQueue\IQueue
{
    private $head = null;
    private $tail = null;
    private $length = 0;

    public function __construct()
    {
    }

    public function size(): int
    {
        return $this->length;
    }

    public function isEmpty(): bool
    {
        return $this->length == 0;
    }

    public function peek(): int
    {
        if ($this->length == 0) {
            throw new RuntimeException("[PANIC - NoSuchElement]");
        }

        return $this->head->data;
    }

    public function enqueue(int $element): void
    {
        $node = new Node($element);

        if ($this->length == 0) {
            $this->head = $node;
        } else {
            $this->tail->next = $node;
        }

        $this->tail = $node;

        $this->length++;
    }

    public function dequeue(): void
    {
        if ($this->length == 0) {
            throw new RuntimeException("[PANIC - NoSuchElement]");
        }

        $target = $this->head;

        if ($this->length == 1) {
            $this->head = null;
            $this->tail = null;
        } else {
            $this->head = $target->next;
        }

        $target->data = null;

        $this->length--;
    }
}

File: java/muse/data_structures/LinkedQueue.java
package muse.data_structures;

import muse.data_structures.interfaces.IQueue;

public class LinkedQueue<E> implements IQueue<E> {
  private static class Node<E> {
    public E data;
    public Node<E> next = null;

    public Node(E element) {
      data = element;
    }
  }

  private Node<E> head = null;
  private Node<E> tail = null;
  private int length = 0;

  @Override
  public int size() {
    return length;
  }

  @Override
  public boolean isEmpty() {
    return length == 0;
  }

  @Override
  public E peek() {
    if (length == 0) {
      throw new RuntimeException("[PANIC - NoSuchElement]");
    }

    return head.data;
  }

  @Override
  @SuppressWarnings("unchecked")
  public void enqueue(E element) {
    Node<E> node = new Node<>(element);

    if (length == 0) {
      head = node;
    } else {
      tail.next = node;
    }

    tail = node;

    length++;
  }

  @Override
  public void dequeue() {
    if (length == 0) {
      throw new RuntimeException("[PANIC - NoSuchElement]");
    }

    Node<E> target = head;

    if (length == 1) {
      head = null;
      tail = null;
    } else {
      head = target.next;
    }

    target.data = null;

    length--;
  }
}
